ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -136,6 +138,50 @@ public boolean setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
     }
   }
 
+  @Override
+  public boolean setReplicaStoppedState(List<ReplicaId> replicaIds, boolean isStopped) {
+    Set<String> replicasToUpdate = new HashSet<>();
+    for (ReplicaId replicaId : replicaIds) {
+      if (!(replicaId instanceof AmbryReplica)) {
+        throw new IllegalArgumentException(
+            "HelixParticipant only works with the AmbryReplica implementation of ReplicaId");
+      }
+      replicasToUpdate.add(replicaId.getPartitionId().toPathString());
+    }
+    boolean setResult;
+    synchronized (helixAdministrationLock) {
+      logger.trace("Getting stopped replicas from instanceConfig");
+      List<String> stoppedListInHelix = getStoppedReplicas();
+      if (stoppedListInHelix == null) {
+        stoppedListInHelix = new ArrayList<>();
+      }
+      Set<String> stoppedListInHelixCopy = new HashSet<>(stoppedListInHelix);
+      if (!isStopped) {
```

Comment:
Firstly, You should call the set `stoppedSet` rather than `stoppedListCopy`.

`removeAll()`, `retainAll()`, `addAll()`, etc. return false if the operation did not change the underlying collection, so you could use that to make this more concise:

```java
stoppedSet = new HashSet<>(stoppedListInHelix);
boolean stoppedSetUpdated = stop ? stoppedSet.addAll(replicasToUpdate) : stoppedSet.removeAll(replicasToUpdate);
return stoppedSetUpdated ? setStoppedReplicas(new ArrayList<>(stoppedSet)) : true;
```

The logging may add a few more lines though. While at it, I don't think we should log these at INFO. This is just the action in response to a request and the public access logging should be sufficient.

Reply:
Done. Thanks for the suggestion.